### DISCUSSION TITLE
Function cached values for aero functions

### DISCUSSION THREAD

**Discussion by jonsberndt** [JSBSim-Team/jsbsim#355] [Support requests]
I see that FGFunctions have been expanded to include caching of aerodynamic values. I'm not sure what drove this, but I am getting a strange error that causes execution to stop in the first pass. Apparently I am getting an FPE when executing aero function code with a vFNative vector that is somehow set to {-inf, 0.0, 0.0). Various example scripts seem to run just fine so perhaps this is a special case. When looking at the function code I noticed this:
```c++
  FGFunction()
    : cached(false), cachedValue(-HUGE_VAL), PropertyManager(nullptr),
      pNode(nullptr), pCopyTo(nullptr) {}
```
I'm wondering why cachedValue is initialized to -HUGE_VAL.

**Comment by jonsberndt** [comment]
To further describe the issue that I am seeing, this works:
```xml
    <function name="aero/force/CDi">
      <description> Induced_drag </description>
      <quotient>
        <pow>
          <value> 2 </value>
          <value> 3.0 </value>
        </pow>   
        <product>
          <value> 2.0 </value>
          <value> 2.0 </value>
        </product>
      </quotient>   
    </function>
```
This does not:
```xml
    <function name="aero/force/CDi">
      <description> Induced_drag </description>
      <quotient>
        <pow>
          <value> 2 </value>
          <value> 3.0 </value>
        </pow>   
        <product>
          <property> aero/qbar-psf </property>   <!-- property here doesn't work -->
          <value> 2.0 </value>
        </product>
      </quotient>   
    </function>
```

**Reply by bcoconni** [reply]
Regarding the first function
```xml
<function name="aero/force/CDi">
  <description> Induced_drag </description>
  <quotient>
    <pow>
      <value> 2 </value>
      <value> 3.0 </value>
    </pow>   
    <product>
      <value> 2.0 </value>
      <value> 2.0 </value>
    </product>
  </quotient>   
</function>
```
It is using constant values only which JSBSim optimizes by pre-computing such functions (see the relevant piece of code below).
https://github.com/JSBSim-Team/jsbsim/blob/22d57cd8849fb80c4ab8fcb3f48c15e20b89f9e0/src/math/FGFunction.cpp#L853-L879
So if everything worked well, JSBSim has (sort of) replaced the code above by
```xml
<property value="2.0"> aero/force/CDi </property>
```
and forced the property to be read only. So, from that perspective, the 2 functions that you have tested are quite differently handled by JSBSim.

**Comment by jonsberndt** [comment]
The problem happens as the simulation is being initialized.

**Comment by jonsberndt** [comment]
I put this same construct in a different, existing, XML file in the aerodynamic portion and it worked just fine.

**Reply by seanmcleod** [reply]
Can you share the complete file which reproduces the error?

I just tried the snippet by adding it to the DRAG section of JSBSim's 737 model without any issues.

**Comment by bcoconni** [comment]
> I see that FGFunctions have been expanded to include caching of aerodynamic values. I'm not sure what drove this, [...]
I'm wondering why cachedValue is initialized to -HUGE_VAL.

Regarding `FGFunction` I can trace the existence of cached values and the usage of `-HUGE_VAL` back to at least Jun 13, 2005 (see below line 36 of commit bd905aa07d - committed by @jonsberndt :wink:). It has been kept like that since then.
https://github.com/JSBSim-Team/jsbsim/blob/bd905aa07d789c1716bde7cef5255c801a48e47b/src/math/FGFunction.cpp#L28-L36
I can only guess that the cached value is initialized to `-HUGE_VAL` to raise an FPE if someone tries to use it before it has been initialized.

Regarding the reason why aerodynamic values are cached, this was committed in May 22, 2016 (commit 18c143a1f81129d by @bcoconni :smile:) and the comment in the code explains that if values are not cached then the aerodynamic values exported by `FGOutput` might differ from the values used to compute the aerodynamic forces.
https://github.com/JSBSim-Team/jsbsim/blob/22d57cd8849fb80c4ab8fcb3f48c15e20b89f9e0/src/models/FGAerodynamics.cpp#L189-L196

**Reply by jonsberndt** [reply]
I vaguely remember writing the caching capability due to something I was seeing in flight controls definitions. Or so I thought. That was a long time ago.

It's a strange issue, and just in this one case. I see the FPE in executables compiled under Cygwin and VC++ both. I will try and make a small example file that illustrates the problem. I'm wondering if there is a very unusual and rare memory leak somewhere. Has a memory test been run on the code lately?

**Reply by bcoconni** [reply]
I know that JSBSim issues a strange error message when no `<aerodynamics>` section is specified in the model. May be you misspelled `aerodynamics` ? Just a wild guess...
> I'm wondering if there is a very unusual and rare memory leak somewhere. Has a memory test been run on the code lately?

The first 5 seconds of the script [Short_S23_3.xml](https://github.com/JSBSim-Team/jsbsim/blob/master/scripts/Short_S23_3.xml) are automatically tested with `valgrind` every time a commit is pushed to the `master` branch.
https://github.com/JSBSim-Team/jsbsim/blob/8819410cc0aee3037de355b0ded0d22c15052df3/.github/workflows/cpp-python-build.yml#L58-L61
This process has already caught a couple of [memory leaks that I inadvertently pushed](https://github.com/JSBSim-Team/jsbsim/runs/1506538412?check_suite_focus=true). Having said that, at the current stage the main virtue of this test is to exist. If we have some new sets of conditions to test then they must certainly be added to our CI workflow.

However, for an FPE, I would rather look for uninitialized variables. Such problems can lead to random crashes and in particular to FPEs. We already had such an issue with `FGPiston` which has been fixed by commit b39803fbc22ad41f29c1cd37. Fortunately, for that particular issue, `valgrind` caught the problem and this is what lead both to the fix and to systemically running `valgrind` in the CI workflow. I am suspecting that we still have uninitialized variables somewhere in the code since we continue to experience random crashes (FPEs by the way) in our test suite. I have yet to find a way to determine where they are (`valgrind` has not found anything to date).

On the other hand, FPEs resulting from uninitialized variables occur seldomly while I understand that your model failed systemically. So we might need to seek something else...
> I will try and make a small example file that illustrates the problem.

That would help greatly ! :+1:
